File: android/app/src/main/java/com/warnabrotha/app/ui/screens/ScanTab.kt
package com.warnabrotha.app.ui.screens

import android.net.Uri
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.CameraAlt
import androidx.compose.material.icons.outlined.CheckCircle
import androidx.compose.material.icons.outlined.ErrorOutline
import androidx.compose.material.icons.outlined.PhotoLibrary
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import coil.request.ImageRequest
import com.warnabrotha.app.data.model.TicketScanResponse
import com.warnabrotha.app.ui.theme.*
import com.warnabrotha.app.ui.viewmodel.ScanState

@Composable
fun ScanTab(
    scanState: ScanState,
    scanImageUri: Uri?,
    scanResult: TicketScanResponse?,
    scanError: String?,
    isLoading: Boolean,
    onTakePhoto: () -> Unit,
    onPickFromLibrary: () -> Unit,
    onSubmitScan: () -> Unit,
    onResetScan: () -> Unit,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier
            .fillMaxSize()
            .background(Background)
            .padding(horizontal = 20.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        when (scanState) {
            ScanState.IDLE -> IdleContent(
                onTakePhoto = onTakePhoto,
                onPickFromLibrary = onPickFromLibrary
            )
            ScanState.PREVIEW -> PreviewContent(
                imageUri = scanImageUri,
                onSubmit = onSubmitScan,
                onRetake = onResetScan
            )
            ScanState.PROCESSING -> ProcessingContent()
            ScanState.SUCCESS -> SuccessContent(
                result = scanResult,
                onScanAnother = onResetScan
            )
            ScanState.ERROR -> ErrorContent(
                error = scanError,
                onTryAgain = onResetScan
            )
        }
    }
}

@Composable
private fun IdleContent(
    onTakePhoto: () -> Unit,
    onPickFromLibrary: () -> Unit
) {
    Spacer(modifier = Modifier.height(80.dp))

    // Camera icon circle
    Box(
        modifier = Modifier
            .size(120.dp)
            .clip(CircleShape)
            .background(GreenOverlay10)
            .border(2.dp, Green500, CircleShape),
        contentAlignment = Alignment.Center
    ) {
        Icon(
            imageVector = Icons.Outlined.CameraAlt,
            contentDescription = "Camera",
            tint = Green500,
            modifier = Modifier.size(48.dp)
        )
    }

    Spacer(modifier = Modifier.height(24.dp))

    Text(
        text = "Scan a Ticket",
        style = MaterialTheme.typography.displayMedium.copy(
            fontFamily = DmSansFamily,
            fontWeight = FontWeight.ExtraBold,
            letterSpacing = (-0.75).sp
        ),
        color = TextPrimary
    )

    Spacer(modifier = Modifier.height(8.dp))

    Text(
        text = "Take a photo of your parking ticket to auto-report TAPS",
        style = MaterialTheme.typography.bodyMedium,
        color = TextSecondary,
        textAlign = TextAlign.Center,
        modifier = Modifier.padding(horizontal = 20.dp)
    )

    Spacer(modifier = Modifier.height(40.dp))

    // Take Photo button
    Button(
        onClick = onTakePhoto,
        modifier = Modifier
            .fillMaxWidth()
            .height(56.dp),
        shape = RoundedCornerShape(24.dp),
        colors = ButtonDefaults.buttonColors(
            containerColor = Green500,
            contentColor = TextOnPrimary
        )
    ) {
        Icon(
            imageVector = Icons.Outlined.CameraAlt,
            contentDescription = null,
            modifier = Modifier.size(20.dp)
        )
        Spacer(modifier = Modifier.width(8.dp))
        Text(
            text = "Take Photo",
            style = MaterialTheme.typography.labelLarge.copy(
                fontFamily = DmSansFamily,
                fontWeight = FontWeight.Bold
            )
        )
    }

    Spacer(modifier = Modifier.height(12.dp))

    // Choose from Library button
    OutlinedButton(
        onClick = onPickFromLibrary,
        modifier = Modifier
            .fillMaxWidth()
            .height(56.dp),
        shape = RoundedCornerShape(24.dp),
        colors = ButtonDefaults.outlinedButtonColors(
            contentColor = Green500
        ),
        border = androidx.compose.foundation.BorderStroke(1.5.dp, Green500)
    ) {
        Icon(
            imageVector = Icons.Outlined.PhotoLibrary,
            contentDescription = null,
            modifier = Modifier.size(20.dp)
        )
        Spacer(modifier = Modifier.width(8.dp))
        Text(
            text = "Choose from Library",
            style = MaterialTheme.typography.labelLarge.copy(
                fontFamily = DmSansFamily,
                fontWeight = FontWeight.Bold
            )
        )
    }
}

@Composable
private fun PreviewContent(
    imageUri: Uri?,
    onSubmit: () -> Unit,
    onRetake: () -> Unit
) {
    Spacer(modifier = Modifier.height(32.dp))

    Text(
        text = "Preview",
        style = MaterialTheme.typography.headlineLarge,
        color = TextPrimary
    )

    Spacer(modifier = Modifier.height(16.dp))

    // Image preview
    if (imageUri != null) {
        AsyncImage(
            model = ImageRequest.Builder(LocalContext.current)
                .data(imageUri)
                .crossfade(true)
                .build(),
            contentDescription = "Ticket preview",
            modifier = Modifier
                .fillMaxWidth()
                .height(300.dp)
                .clip(RoundedCornerShape(16.dp))
                .border(1.dp, Border, RoundedCornerShape(16.dp)),
            contentScale = ContentScale.Fit
        )
    }

    Spacer(modifier = Modifier.height(24.dp))

    // Submit button
    Button(
        onClick = onSubmit,
        modifier = Modifier
            .fillMaxWidth()
            .height(56.dp),
        shape = RoundedCornerShape(24.dp),
        colors = ButtonDefaults.buttonColors(
            containerColor = Green500,
            contentColor = TextOnPrimary
        )
    ) {
        Text(
            text = "Submit Ticket",
            style = MaterialTheme.typography.labelLarge.copy(
                fontFamily = DmSansFamily,
                fontWeight = FontWeight.Bold
            )
        )
    }

    Spacer(modifier = Modifier.height(12.dp))

    // Retake button
    TextButton(
        onClick = onRetake,
        modifier = Modifier.fillMaxWidth()
    ) {
        Text(
            text = "Retake",
            style = MaterialTheme.typography.labelLarge,
            color = TextSecondary
        )
    }
}

@Composable
private fun ProcessingContent() {
    Spacer(modifier = Modifier.weight(1f))

    CircularProgressIndicator(
        color = Green500,
        modifier = Modifier.size(48.dp),
        strokeWidth = 4.dp
    )

    Spacer(modifier = Modifier.height(24.dp))

    Text(
        text = "Reading ticket...",
        style = MaterialTheme.typography.headlineMedium,
        color = TextPrimary
    )

    Spacer(modifier = Modifier.height(8.dp))

    Text(
        text = "Extracting date, time, and location",
        style = MaterialTheme.typography.bodyMedium,
        color = TextSecondary
    )

    Spacer(modifier = Modifier.weight(1f))
}

@Composable
private fun SuccessContent(
    result: TicketScanResponse?,
    onScanAnother: () -> Unit
) {
    Spacer(modifier = Modifier.height(48.dp))

    // Success icon
    Box(
        modifier = Modifier
            .size(80.dp)
            .clip(CircleShape)
            .background(GreenOverlay10),
        contentAlignment = Alignment.Center
    ) {
        Icon(
            imageVector = Icons.Outlined.CheckCircle,
            contentDescription = "Success",
            tint = Green500,
            modifier = Modifier.size(48.dp)
        )
    }

    Spacer(modifier = Modifier.height(24.dp))

    if (result != null) {
        // Extracted data card
        Surface(
            modifier = Modifier.fillMaxWidth(),
            shape = RoundedCornerShape(16.dp),
            color = Surface,
            border = androidx.compose.foundation.BorderStroke(1.dp, Border)
        ) {
            Column(modifier = Modifier.padding(20.dp)) {
                Text(
                    text = "Ticket Details",
                    style = MaterialTheme.typography.headlineSmall,
                    color = TextPrimary
                )

                Spacer(modifier = Modifier.height(12.dp))

                result.ticketDate?.let { date ->
                    DetailRow(label = "Date", value = date)
                }
                result.ticketTime?.let { time ->
                    DetailRow(label = "Time", value = time)
                }
                result.ticketLocation?.let { location ->
                    DetailRow(label = "Location", value = location)
                }
                result.mappedLotName?.let { lotName ->
                    DetailRow(label = "Mapped to", value = lotName)
                }
            }
        }

        Spacer(modifier = Modifier.height(20.dp))

        // Result message
        if (result.isRecent && result.sightingId != null) {
            Surface(
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(16.dp),
                color = GreenOverlay10
            ) {
                Column(
                    modifier = Modifier.padding(20.dp),
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    Text(
                        text = "TAPS report created!",
                        style = MaterialTheme.typography.headlineSmall,
                        color = Green600
                    )
                    Text(
                        text = "${result.usersNotified} users notified",
                        style = MaterialTheme.typography.bodyMedium,
                        color = Green600
                    )
                }
            }
        } else if (result.sightingId != null) {
            Surface(
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(16.dp),
                color = CardBackgroundMuted
            ) {
                Column(
                    modifier = Modifier.padding(20.dp),
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    Text(
                        text = "Ticket recorded",
                        style = MaterialTheme.typography.headlineSmall,
                        color = TextPrimary
                    )
                    Text(
                        text = "Too old for a live report",
                        style = MaterialTheme.typography.bodyMedium,
                        color = TextSecondary
                    )
                }
            }
        } else if (result.mappedLotId == null && result.ticketLocation != null) {
            Surface(
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(16.dp),
                color = RedOverlay10
            ) {
                Column(
                    modifier = Modifier.padding(20.dp),
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    Text(
                        text = "Location not recognized",
                        style = MaterialTheme.typography.headlineSmall,
                        color = Red500
                    )
                    Text(
                        text = "Couldn't create a report for this location",
                        style = MaterialTheme.typography.bodyMedium,
                        color = TextSecondary
                    )
                }
            }
        }
    }

    Spacer(modifier = Modifier.height(32.dp))

    Button(
        onClick = onScanAnother,
        modifier = Modifier
            .fillMaxWidth()
            .height(56.dp),
        shape = RoundedCornerShape(24.dp),
        colors = ButtonDefaults.buttonColors(
            containerColor = Green500,
            contentColor = TextOnPrimary
        )
    ) {
        Text(
            text = "Scan Another",
            style = MaterialTheme.typography.labelLarge.copy(
                fontFamily = DmSansFamily,
                fontWeight = FontWeight.Bold
            )
        )
    }
}

@Composable
private fun DetailRow(label: String, value: String) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 4.dp),
        horizontalArrangement = Arrangement.SpaceBetween
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.bodyMedium,
            color = TextSecondary
        )
        Text(
            text = value,
            style = MaterialTheme.typography.titleMedium,
            color = TextPrimary
        )
    }
}

@Composable
private fun ErrorContent(
    error: String?,
    onTryAgain: () -> Unit
) {
    Spacer(modifier = Modifier.weight(1f))

    // Error icon
    Box(
        modifier = Modifier
            .size(80.dp)
            .clip(CircleShape)
            .background(RedOverlay10),
        contentAlignment = Alignment.Center
    ) {
        Icon(
            imageVector = Icons.Outlined.ErrorOutline,
            contentDescription = "Error",
            tint = Red500,
            modifier = Modifier.size(48.dp)
        )
    }

    Spacer(modifier = Modifier.height(24.dp))

    Text(
        text = "Could not read ticket",
        style = MaterialTheme.typography.headlineLarge,
        color = TextPrimary,
        textAlign = TextAlign.Center
    )

    Spacer(modifier = Modifier.height(8.dp))

    Text(
        text = error ?: "Make sure the photo is clear and shows a UC Davis parking ticket.",
        style = MaterialTheme.typography.bodyMedium,
        color = TextSecondary,
        textAlign = TextAlign.Center,
        modifier = Modifier.padding(horizontal = 20.dp)
    )

    Spacer(modifier = Modifier.height(32.dp))

    Button(
        onClick = onTryAgain,
        modifier = Modifier
            .fillMaxWidth()
            .height(56.dp),
        shape = RoundedCornerShape(24.dp),
        colors = ButtonDefaults.buttonColors(
            containerColor = Green500,
            contentColor = TextOnPrimary
        )
    ) {
        Text(
            text = "Try Again",
            style = MaterialTheme.typography.labelLarge.copy(
                fontFamily = DmSansFamily,
                fontWeight = FontWeight.Bold
            )
        )
    }

    Spacer(modifier = Modifier.weight(1f))
}

File: warnabrotha/warnabrotha/Views/ScanTab.swift
//
//  ScanTab.swift
//  TapOut
//
//  Ticket scanning — Records history list + camera/gallery capture → OCR upload → result display.
//

import SwiftUI
import PhotosUI

struct ScanTab: View {
    @ObservedObject var viewModel: AppViewModel

    var body: some View {
        VStack(spacing: 0) {
            ScanHeader(viewModel: viewModel)

            switch viewModel.scanSubTab {
            case .records:
                TicketHistoryView(viewModel: viewModel)
            case .scanner:
                scannerContent
            }
        }
        .background(AppColors.background)
    }

    @ViewBuilder
    private var scannerContent: some View {
        switch viewModel.scanState {
        case .idle:
            ScanIdleView(viewModel: viewModel)
        case .preview:
            ScanPreviewView(viewModel: viewModel)
        case .processing:
            ScanProcessingView()
        case .success:
            ScanSuccessView(viewModel: viewModel)
        case .error:
            ScanErrorView(viewModel: viewModel)
        }
    }
}

// MARK: - Shared Header

private struct ScanHeader: View {
    @ObservedObject var viewModel: AppViewModel

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            // Branding + title
            VStack(alignment: .leading, spacing: 4) {
                (Text("Tap")
                    .foregroundColor(AppColors.textPrimary)
                + Text("Out")
                    .foregroundColor(AppColors.accent))
                    .appFont(size: 11, weight: .bold)

                Text("Scan Ticket")
                    .appFont(size: 30, weight: .heavy)
                    .foregroundColor(AppColors.textPrimary)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, 24)
            .padding(.top, 12)
            .padding(.bottom, 12)

            // Underline tab bar
            HStack(spacing: 0) {
                ScanTabButton(label: "Scanner", isActive: viewModel.scanSubTab == .scanner) {
                    viewModel.scanSubTab = .scanner
                }
                ScanTabButton(label: "Records", isActive: viewModel.scanSubTab == .records) {
                    viewModel.scanSubTab = .records
                }
            }
            .overlay(alignment: .bottom) {
                Rectangle()
                    .fill(AppColors.border)
                    .frame(height: 1)
            }
        }
    }
}

private struct ScanTabButton: View {
    let label: String
    let isActive: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            VStack(spacing: 0) {
                Text(label)
                    .appFont(size: 15, weight: .bold)
                    .foregroundColor(isActive ? AppColors.accent : AppColors.textMuted)
                    .padding(.horizontal, 24)
                    .padding(.vertical, 10)

                Rectangle()
                    .fill(isActive ? AppColors.accent : Color.clear)
                    .frame(height: 2)
            }
        }
        .buttonStyle(PlainButtonStyle())
    }
}

// MARK: - Ticket History (Records Tab)

private struct TicketHistoryView: View {
    @ObservedObject var viewModel: AppViewModel

    var body: some View {
        if viewModel.ticketHistory.isEmpty {
            VStack(spacing: 12) {
                Image(systemName: "doc.text.magnifyingglass")
                    .font(.system(size: 40, weight: .light))
                    .foregroundColor(AppColors.textMuted)

                Text("No tickets scanned yet")
                    .appFont(size: 16, weight: .medium)
                    .foregroundColor(AppColors.textSecondary)

                Text("Scan a parking ticket to see it here.")
                    .appFont(size: 13)
                    .foregroundColor(AppColors.textMuted)
                    .multilineTextAlignment(.center)
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .padding(.top, 80)
        } else {
            ScrollView {
                LazyVStack(spacing: 12) {
                    ForEach(viewModel.ticketHistory) { entry in
                        TicketHistoryRow(entry: entry)
                    }
                }
                .padding(.horizontal, 16)
                .padding(.top, 16)
                .padding(.bottom, 32)
            }
        }
    }
}

private struct TicketHistoryRow: View {
    let entry: TicketHistoryEntry

    private var displayLot: String {
        (entry.lotCode ?? entry.lotName ?? entry.ticketLocation ?? "Unknown").uppercased()
    }

    private var displayDate: String {
        if let d = entry.ticketDate { return d }
        let f = DateFormatter(); f.dateStyle = .short; f.timeStyle = .none
        return f.string(from: entry.scannedAt)
    }

    private var displayTime: String {
        if let t = entry.ticketTime { return t }
        let f = DateFormatter(); f.dateStyle = .none; f.timeStyle = .short
        return f.string(from: entry.scannedAt)
    }

    var body: some View {
        HStack(alignment: .center, spacing: 16) {
            // Left: time above, lot code below — mirrors FeedCardView
            VStack(alignment: .leading, spacing: 4) {
                Text(displayTime)
                    .appFont(size: 11, weight: .bold)
                    .foregroundColor(AppColors.accent)

                Text(displayLot)
                    .appFont(size: 22, weight: .bold)
                    .foregroundColor(AppColors.textPrimary)
                    .lineLimit(1)
            }

            Spacer()

            // Right: date, then price below
            VStack(alignment: .trailing, spacing: 4) {
                Text(displayDate)
                    .appFont(size: 14, weight: .bold)
                    .foregroundColor(AppColors.textMuted)
                if let amount = entry.ticketAmount {
                    Text(amount)
                        .appFont(size: 14, weight: .bold)
                        .foregroundColor(AppColors.textMuted)
                }
            }
        }
        .padding(.horizontal, 20)
        .padding(.vertical, 20)
        .background(
            RoundedRectangle(cornerRadius: 16)
                .fill(AppColors.cardBackground)
        )
        .shadow(color: .black.opacity(0.04), radius: 4, y: 2)
    }
}

// MARK: - Idle State

private struct ScanIdleView: View {
    @ObservedObject var viewModel: AppViewModel
    @State private var showCamera = false
    @State private var showPhotoPicker = false
    @State private var selectedPhotoItem: PhotosPickerItem? = nil

    var body: some View {
        VStack(spacing: 0) {
            Spacer()

            VStack(spacing: 32) {
                // Camera icon
                ZStack {
                    Circle()
                        .stroke(AppColors.border, lineWidth: 2)
                        .frame(width: 120, height: 120)

                    Image(systemName: "doc.viewfinder")
                        .font(.system(size: 48, weight: .light))
                        .foregroundColor(AppColors.accent)
                }

                VStack(spacing: 8) {
                    Text("Scan a Ticket")
                        .displayFont(size: 24)
                        .foregroundColor(AppColors.textPrimary)

                    Text("Take a photo of your parking ticket to auto-report TAPS")
                        .appFont(size: 14)
                        .foregroundColor(AppColors.textSecondary)
                        .multilineTextAlignment(.center)
                        .padding(.horizontal, 40)
                }

                VStack(spacing: 12) {
                    // Take Photo button
                    Button {
                        showCamera = true
                    } label: {
                        HStack(spacing: 8) {
                            Image(systemName: "camera.fill")
                                .font(.system(size: 18, weight: .semibold))
                            Text("Take Photo")
                                .appFont(size: 16, weight: .bold)
                        }
                        .foregroundColor(.white)
                        .frame(maxWidth: .infinity)
                        .frame(height: 56)
                        .background(
                            RoundedRectangle(cornerRadius: 16)
                                .fill(AppColors.accent)
                        )
                    }
                    .buttonStyle(PlainButtonStyle())

                    // Choose from Library button
                    PhotosPicker(
                        selection: $selectedPhotoItem,
                        matching: .images,
                        photoLibrary: .shared()
                    ) {
                        HStack(spacing: 8) {
                            Image(systemName: "photo.on.rectangle")
                                .font(.system(size: 18, weight: .semibold))
                            Text("Choose from Library")
                                .appFont(size: 16, weight: .bold)
                        }
                        .foregroundColor(AppColors.accent)
                        .frame(maxWidth: .infinity)
                        .frame(height: 56)
                        .background(
                            RoundedRectangle(cornerRadius: 16)
                                .stroke(AppColors.accent, lineWidth: 2)
                        )
                    }
                    .buttonStyle(PlainButtonStyle())
                    .onChange(of: selectedPhotoItem) { _, newItem in
                        guard let newItem else { return }
                        Task {
                            if let data = try? await newItem.loadTransferable(type: Data.self) {
                                viewModel.selectScanImage(data)
                            }
                        }
                    }
                }
                .padding(.horizontal, 24)
            }

            Spacer()
        }
        .fullScreenCover(isPresented: $showCamera) {
            CameraView { imageData in
                viewModel.selectScanImage(imageData)
            }
            .ignoresSafeArea()
        }
    }
}

// MARK: - Preview State

private struct ScanPreviewView: View {
    @ObservedObject var viewModel: AppViewModel

    var body: some View {
        VStack(spacing: 0) {
            Spacer()

            if let data = viewModel.scanImageData, let uiImage = UIImage(data: data) {
                Image(uiImage: uiImage)
                    .resizable()
                    .scaledToFit()
                    .frame(maxHeight: 300)
                    .clipShape(RoundedRectangle(cornerRadius: 16))
                    .shadow(color: .black.opacity(0.1), radius: 8, y: 4)
                    .padding(.horizontal, 24)
            }

            Spacer()

            VStack(spacing: 12) {
                PrimaryButton(
                    title: "Submit Ticket",
                    icon: "arrow.up.circle.fill",
                    action: {
                        Task { await viewModel.submitTicketScan() }
                    }
                )

                Button {
                    viewModel.resetScan()
                } label: {
                    Text("Retake")
                        .appFont(size: 16, weight: .semibold)
                        .foregroundColor(AppColors.textSecondary)
                }
                .buttonStyle(PlainButtonStyle())
            }
            .padding(.horizontal, 24)
            .padding(.bottom, 32)
        }
    }
}

// MARK: - Processing State

private struct ScanProcessingView: View {
    var body: some View {
        VStack(spacing: 16) {
            Spacer()

            ProgressView()
                .scaleEffect(1.5)
                .tint(AppColors.accent)

            Text("Reading ticket...")
                .displayFont(size: 20)
                .foregroundColor(AppColors.textPrimary)

            Text("Extracting date, time, and location")
                .appFont(size: 14)
                .foregroundColor(AppColors.textSecondary)

            Spacer()
        }
    }
}

// MARK: - Success State

private struct ScanSuccessView: View {
    @ObservedObject var viewModel: AppViewModel

    var body: some View {
        ScrollView {
            VStack(spacing: 24) {
                Spacer(minLength: 32)

                // Success icon
                ZStack {
                    Circle()
                        .fill(AppColors.accent.opacity(0.1))
                        .frame(width: 80, height: 80)

                    Image(systemName: "checkmark.circle.fill")
                        .font(.system(size: 40))
                        .foregroundColor(AppColors.accent)
                }

                if let result = viewModel.scanResult {
                    // Ticket Details card
                    VStack(alignment: .leading, spacing: 16) {
                        Text("TICKET DETAILS")
                            .appFont(size: 10, weight: .bold)
                            .tracking(1)
                            .foregroundColor(AppColors.textMuted)

                        VStack(spacing: 12) {
                            if let date = result.ticketDate {
                                DetailRow(label: "Date", value: date, icon: "calendar")
                            }
                            if let time = result.ticketTime {
                                DetailRow(label: "Time", value: time, icon: "clock")
                            }
                            if let location = result.ticketLocation {
                                DetailRow(label: "Location", value: location, icon: "mappin.circle")
                            }
                            if let lotName = result.mappedLotName {
                                DetailRow(label: "Mapped Lot", value: lotName, icon: "car.fill")
                            }
                        }
                    }
                    .padding(20)
                    .cardStyle(cornerRadius: 16)
                    .padding(.horizontal, 24)

                    // Result message
                    resultMessageView(result: result)
                        .padding(.horizontal, 24)
                }

                PrimaryButton(
                    title: "Scan Another",
                    icon: "doc.viewfinder",
                    action: { viewModel.resetScan() }
                )
                .padding(.horizontal, 24)

                Button {
                    viewModel.resetScan()
                    viewModel.scanSubTab = .records
                } label: {
                    Text("View History")
                        .appFont(size: 14, weight: .semibold)
                        .foregroundColor(AppColors.accent)
                }
                .buttonStyle(PlainButtonStyle())

                Spacer(minLength: 32)
            }
        }
    }

    @ViewBuilder
    private func resultMessageView(result: TicketScanResponse) -> some View {
        if result.isRecent && result.sightingId != nil {
            HStack(spacing: 8) {
                Image(systemName: "checkmark.circle.fill")
                    .foregroundColor(AppColors.accent)
                Text("TAPS report created! \(result.usersNotified) users notified.")
                    .appFont(size: 14, weight: .medium)
                    .foregroundColor(AppColors.accent)
            }
            .padding(16)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(
                RoundedRectangle(cornerRadius: 16)
                    .fill(AppColors.accent.opacity(0.1))
            )
        } else if !result.isRecent {
            HStack(spacing: 8) {
                Image(systemName: "clock.fill")
                    .foregroundColor(AppColors.textMuted)
                Text("Ticket recorded — Too old for a live report")
                    .appFont(size: 14, weight: .medium)
                    .foregroundColor(AppColors.textMuted)
            }
            .padding(16)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(
                RoundedRectangle(cornerRadius: 16)
                    .fill(AppColors.background)
            )
        } else if result.mappedLotId == nil {
            HStack(spacing: 8) {
                Image(systemName: "exclamationmark.triangle.fill")
                    .foregroundColor(AppColors.danger)
                Text("Location not recognized")
                    .appFont(size: 14, weight: .medium)
                    .foregroundColor(AppColors.danger)
            }
            .padding(16)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(
                RoundedRectangle(cornerRadius: 16)
                    .fill(AppColors.dangerLight)
            )
        }
    }
}

// MARK: - Error State

private struct ScanErrorView: View {
    @ObservedObject var viewModel: AppViewModel

    var body: some View {
        VStack(spacing: 24) {
            Spacer()

            ZStack {
                Circle()
                    .fill(AppColors.dangerLight)
                    .frame(width: 80, height: 80)

                Image(systemName: "xmark.circle.fill")
                    .font(.system(size: 40))
                    .foregroundColor(AppColors.danger)
            }

            VStack(spacing: 8) {
                Text("Could not read ticket")
                    .displayFont(size: 20)
                    .foregroundColor(AppColors.textPrimary)

                Text(viewModel.scanError ?? "Make sure the photo is clear and shows a UC Davis parking ticket.")
                    .appFont(size: 14)
                    .foregroundColor(AppColors.textSecondary)
                    .multilineTextAlignment(.center)
                    .padding(.horizontal, 40)
            }

            PrimaryButton(
                title: "Try Again",
                icon: "arrow.counterclockwise",
                action: { viewModel.resetScan() }
            )
            .padding(.horizontal, 24)

            Spacer()
        }
    }
}

// MARK: - Detail Row

private struct DetailRow: View {
    let label: String
    let value: String
    let icon: String

    var body: some View {
        HStack(spacing: 12) {
            Image(systemName: icon)
                .font(.system(size: 16))
                .foregroundColor(AppColors.accent)
                .frame(width: 24)

            VStack(alignment: .leading, spacing: 2) {
                Text(label)
                    .appFont(size: 10, weight: .bold)
                    .foregroundColor(AppColors.textMuted)
                    .textCase(.uppercase)
                Text(value)
                    .appFont(size: 14, weight: .medium)
                    .foregroundColor(AppColors.textPrimary)
            }

            Spacer()
        }
    }
}

// MARK: - Camera View (UIImagePickerController wrapper)

struct CameraView: UIViewControllerRepresentable {
    let onCapture: (Data) -> Void
    @Environment(\.dismiss) private var dismiss

    func makeUIViewController(context: Context) -> UIImagePickerController {
        let picker = UIImagePickerController()
        picker.sourceType = .camera
        picker.delegate = context.coordinator
        return picker
    }

    func updateUIViewController(_ uiViewController: UIImagePickerController, context: Context) {}

    func makeCoordinator() -> Coordinator {
        Coordinator(onCapture: onCapture, dismiss: dismiss)
    }

    class Coordinator: NSObject, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
        let onCapture: (Data) -> Void
        let dismiss: DismissAction

        init(onCapture: @escaping (Data) -> Void, dismiss: DismissAction) {
            self.onCapture = onCapture
            self.dismiss = dismiss
        }

        func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey: Any]) {
            if let image = info[.originalImage] as? UIImage,
               let data = image.jpegData(compressionQuality: 0.8) {
                onCapture(data)
            }
            dismiss()
        }

        func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
            dismiss()
        }
    }
}

#Preview {
    ScanTab(viewModel: AppViewModel())
}
